### DISCUSSION TITLE
Documentation of Standard Patterns

### DISCUSSION THREAD

**Discussion by LeStarch** [nasa/fprime#2732] [Call for Comments]
We are considering the idea of publishing standard patterns as part of F´ documentation.  Thus far, we have come up with the following list.

Please comment and let us know what you'd like to see documented!

1. Implementation patterns (beginner)  - Reference + Example
    - [x] Get/request pattern
    - [x] Callback pattern
    - [ ] Memory allocation pattern -- SUBMITTED
    - [ ]  Buffer manager pattern -- SUBMITTED
    - [x]  Sync cancel 
    - [ ] Internal interfaces -- SUBMITTED
    - [x] Parallel ports
    - [x] Health
    - [ ] State machines -- SUBMITTED
2. Component Patterns: (beginner) - Reference + Example
    1. Central store -- ??
    2. Hub Pattern - Exists
    3. Manager/Worker pattern - Developed
    4. Rate group pattern - Exists.  Improve?
    5. Mode management patterns   -- SUBMITTED
    6. Interrupt handling  -- SUBMITTED
    7. Resource manager -- ??
3. Interface patterns (intermediate) - How To
    1. Generic interfaces  - review guide  -- SUBMITTED
    2. Time/Tick 
    3. Byte stream driver
    4. I2C  -- SUBMITTED
    5. SPI
    6. GPIO
    7. PWM
    8. Watchdog
4. Topology patterns (intermediate) - Decide by Item
    1. Custom configuration -- SUBMITTED
    2. Reused component instances -- ??
    3. Sub topology  -- SUBMITTED
    4. Application/Driver/Manager pattern -- exists
    5. Fault protection  -- SUBMITTED
    6. Sequencer pooling  -- SUBMITTED
    7. Assert -- ??
    8. logger -- ??
    10. Security  -- SUBMITTED
    11. Simulation deployments
5. Integration patterns (advanced) - How To
    1. OS -- SUBMITTED
    2. Hardware Platform
    3. New GDS  -- SUBMITTED
    4. Radio -- SUBMITTED
    5. Framing protocol -- SUBMITTED
    6. Vendor library -- in progress
    7. Other framework 
    8. Ground processing script -- SUBMITTED
6. Modeling and build tool patterns (advanced) - How Guides
    1. Custom autocoder -- SUBMITTED
    2. Custom model validation 
    3. Custom model tool

**Comment by SterlingPeet** [comment]
Command registration and Dispatch as implemented in F Prime is often confusing for newcomers to the framework.

Telemetry handling can be equally confusing, but is generally easier to mimic and thus get it to work.

**Reply by SterlingPeet** [reply]
BTW, I am excited by the prospect of this documentation becoming available, I have had to explain a lot of it in situ for students.  It would be great to be able to index into pattern documentation for learning purposes.

**Reply by LeStarch** [reply]
@SterlingPeet Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Comment by ReggieMarr** [comment]
This is excellent, also fwiw I think the gold standard to strive for is documentation somewhat similar to what's provided by [0mq](https://zguide.zeromq.org/docs/chapter4/). Admittedly it's a massive amount of work but worth keeping in mind.

**Reply by LeStarch** [reply]
Will do!

**Comment by thomas-bc** [comment]
Creating a thread to throw in ideas so as not to spam the main thread 🧵 ⬇️

**Reply by thomas-bc** [reply]
- FPP [Pattern Graph specifiers](https://nasa.github.io/fpp/fpp-users-guide.html#Defining-Topologies_Connection-Graphs_Pattern-Graph-Specifiers) and how to manually define one, say for a ModeManager component

**Reply by ReggieMarr** [reply]
speaking of threads, are there any examples of how to utilize thread virtualization or create custom threads os/task

**Reply by thomas-bc** [reply]
- Stop using `PRIVATE:` for UTs and use the test Friend class for whitebox testing instead

**Reply by LeStarch** [reply]
> * FPP [Pattern Graph specifiers](https://nasa.github.io/fpp/fpp-users-guide.html#Defining-Topologies_Connection-Graphs_Pattern-Graph-Specifiers) and how to manually define one, say for a ModeManager component

Custom pattern-graph specifiers is not possible in FPP.  They'd have to be added to the FPP language itself.

**Reply by LeStarch** [reply]
@ReggieMarr would you mind creating "how to do custom threads" as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Reply by ReggieMarr** [reply]
Sure thing !

**Comment by SiboVG** [comment]
Hey! It would also be great to have an overview of the available built-in FPP types and their use case. I'm referring to something similar to the [C++ documentation](https://nasa.github.io/fprime/UsersGuide/api/c++/html/namespaces.html), but for FPP.

For example, in the Led Blinker workshop, you learn to add some ports to `Led.fpp`:
```fpp
@ Port receiving calls from the rate group
async input port run: Svc.Sched

@ Port sending calls to the GPIO driver
output port gpioSet: Drv.GpioWrite
```

But I have no idea what other FPP port types exist or what their use case is.

**Reply by LeStarch** [reply]
I really like this one!

@SiboVG Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Reply by SiboVG** [reply]
> I really like this one!
> 
> @SiboVG Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)? One for each pattern suggested. This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

@LeStarch, sure, sorry for the late reply! I've posted it [here](https://github.com/nasa/fprime/discussions/3365). Hope it's sufficiently clear.

**Comment by kevin-f-ortega** [comment]
Here are some ideas I have

- [ ] How to add a new toolchain
- [ ] How to integrate C/C++ libraries with F Prime
- [ ] How to port to a new operating system 
- [ ] How to add CI
- [ ] How to structure your code to be reusable by other projects
- [ ] How to write drivers using the Interface paradigm

**Reply by LeStarch** [reply]
@kevin-f-ortega Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Comment by LeStarch** [comment]
Some other ideas that came in via PR #3040:

1. Document the bytetreamdriver pattern
2. Document the buffer allocation/ownership pattern.

**Reply by LeStarch** [reply]
@LeStarch Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Reply by LeStarch** [reply]
Submitted, thank you!

**Comment by SterlingPeet** [comment]
@zeroping may have some more recent experience with what would be helpful tutorials or cookbook recipes to help with onboarding developers

**Reply by LeStarch** [reply]
Recent experience or not, I'd love to hear @zeroping's suggestions!

**Reply by zeroping** [reply]
I've recently watched a few new students try to wrap their heads around making a manager component that talks to an existing hardware driver component (for example, something that manages configuration of a device connected via UART).  Having a guide for things like that would be helpful, since "make the manager component for X" is otherwise a good place to start learning with a real system.

I think the best documentation would be something concrete that someone could follow along with, not just a wall of words that aren't as actionable. Imagine something between demo code and a tutorial, but where the user can follow along the steps, but can apply them to their own use case in parallel with the tutorial.

For example, imagine we had a well-implemented manager for some super common SPI-connected accelerometer, some hardware that anyone could buy from Amazon for $5 if needed. Then have a tutorial that explains:

- How the design decisions were made ("we wanted a queued port here because we may not get a response from this SPI device immediately")
- How the new component is connected to a driver component ("We need a buffer manager here because that's how we get from the driver component")
- How the commands, telemetry, and events, and parameters are chosen ("We want this data to be downlinked through an event rather than a telemetry channel because we very rarely expect it to change at runtime")
- How the FPP was written to support that
- How the code gets written (including some ideas for dealing with places where we need to wait for the hardware to provide data back)
- How to write tests for something that interfaces with real hardware ("To test with without real hardware connected, we can write a  mock component that can stand in for the SPI driver, but responds to SPI reads and writes as if it were an accelerometer", or "We have a HITL test to exercise all of the functionality when an accelerometer is connected to a Raspberry Pi")

Maybe also pair with a cookbook recipe for "manager component that interfaces with an SPI device".

So, that new user gets reference code, plus the reasoning behind it, plus links into the rest of the documentation for larger concepts like port types, plus auto-generated code they can start from. Super actionable when they have a device datasheet or user manual and need to get started.

Hopefully that made some sense.

**Reply by zeroping** [reply]
Another approach to answer your question:

Take any component in the FPrime tree - it's probably there because the core FPrime people needed it to solve a real problem, and made a clean generalized solution. For example, how about [BufferLogger](https://github.com/nasa/fprime/tree/devel/Svc/BufferLogger)? It exists to solve a problem, but I'm not sure that's documented anywhere. Someone looking at what it does might not be able to work that out, and would have a hard time finding it if they needed it. So just a quick "We needed to do X on a project, so this component helped to glue Y to Z. We've also seen it used to solve this other problem A" would be great.

As a concrete example, there was a [discussion](https://github.com/nasa/fprime/discussions/1881) that is a great motivation for [ComSplitter](https://github.com/nasa/fprime/tree/devel/Svc/ComSplitter). Reading through the list of components and going "what all is ComSplitter good for?" could yield a list of common patterns that might be obvious in hindsight, but confusing at first.

**Reply by zeroping** [reply]
One last comment: I think fprime-visual could use some love (I know that's been said elsewhere too). 

If you want to document patterns, a good graph could go a long way to quickly explain how the parts are connected, and right now, fprime-visual won't _quite_ give you a pretty graph without lots of effort. Generating readable component interconnection graphs, especially for cookie-cutter-generated recipes, could be powerful for someone trying to figure out what's what in an example.

**Reply by LeStarch** [reply]
@zeroping  Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Comment by LeStarch** [comment]
https://github.com/nasa/fprime/issues/1844

**Comment by LeStarch** [comment]
https://github.com/nasa/fprime/issues/1688

**Reply by LeStarch** [reply]
Already submitted.

**Comment by LeStarch** [comment]
https://github.com/nasa/fprime/issues/1487

**Comment by LeStarch** [comment]
https://github.com/nasa/fprime/issues/1251

**Comment by LeStarch** [comment]
https://github.com/nasa/fprime/issues/2689

**Reply by LeStarch** [reply]
Done.

**Comment by elliotfayman** [comment]
[#3056](https://github.com/nasa/fprime/issues/3056)

**Reply by LeStarch** [reply]
@elliotfayman Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)?  One for each pattern suggested.  This will allow you and others to [vote on desired patterns](https://github.com/nasa/fprime/discussions/3298)!

**Reply by LeStarch** [reply]
nevermind.  This one is already assigned and in the works.

**Comment by ronaldo-monteiro** [comment]
I would like to see the use of AI in code improvements for the project, especially in Python.

**Reply by LeStarch** [reply]
Not sure how this applies to F Prime development as F Prime is primarily C++. However, feel free to elaborate in a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)

**Comment by kevin-f-ortega** [comment]
Document Patches too?

**Reply by LeStarch** [reply]
@kevin-f-ortega [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)

**Comment by Joshua-Anderson** [comment]
I'd recommend writing a section on fprime unit and integration testing patterns and best practices. - I'm not sure there's any comprehensive guides to fprime testing outside of the LED example.

**Reply by LeStarch** [reply]
@Joshua-Anderson  Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)? One for each pattern suggested. This will allow you and others to[vote](https://github.com/nasa/fprime/discussions/3298)!

**Comment by zimri-leisher** [comment]
Quick note--could add documentation to explain what the pattern is for a custom decoding plugin to the GDS

**Reply by LeStarch** [reply]
@zimri-leisher Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)? One for each pattern suggested. This will allow you and others to [vote](https://github.com/nasa/fprime/discussions/3298)!

**Comment by timcanham** [comment]
###  Some Data Processing ideas

1. Pipelining - (e.g. image processing) - A -> B -> C - show with `Fw::Buffer` lifecycle referenced earlier
2. Decoration - (e.g. image processing - similar to pipeling, for example processor on `DpWriter`) - A->B^C^->D
3. Adapter - A->B (convert) ->C

**Reply by LeStarch** [reply]
@timcanham Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)? One for each pattern suggested. This will allow you and others to [vote](https://github.com/nasa/fprime/discussions/3298)!

**Comment by matt392code** [comment]


**Reply by thomas-bc** [reply]
@matt392code this discussion was set up to collect **feedback** from the community on the patterns we would like to see documented. If you would like to contribute these diagrams, please first ensure they are accurate, and then contribute them to a relevant space (new Discussion, PR etc.)

**Comment by matt392code** [comment]
Sorry about that.
Put it into the Discussion > Show and Tell so that it can be reviewed by the Community.

**Reply by thomas-bc** [reply]
Could you please group them all in a single discussion instead of creating multiple discussions? Thanks!

**Reply by matt392code** [reply]
Will do.  Sorry about that.

**Reply by matt392code** [reply]
Put 2 diagrams in one discussion thread.

**Comment by kubiak-jpl** [comment]
Managing multiple deployments in a single repository. Most projects I've worked on tend to have more than one deployment, usually a flight deployment, but also one or more related test deployment. What is the canonical way to share code and partial topologies between these deployments? Is it one CMake project or multiple? Where do the config files go?

**Reply by kevin-f-ortega** [reply]
Yes, good to know about this. Especially, how to configure the `settings.ini` file when you have different configs for each deployment.

**Reply by LeStarch** [reply]
@kubiak-jpl Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)? One for each pattern suggested. This will allow you and others to [vote](https://github.com/nasa/fprime/discussions/3298)!

**Comment by kevin-f-ortega** [comment]
It would be good to document the delegate pattern and how to use it (especially when writing OSAL).

**Reply by LeStarch** [reply]
@kevin-f-ortega Would you mind creating this as a [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)? One for each pattern suggested. This will allow you and others to [vote](https://github.com/nasa/fprime/discussions/3298)!

Probably "how to write an OSAL implementation" too

**Comment by LeStarch** [comment]
I am closing this discussion. Submit requests through our formal request. [formal request](https://github.com/nasa/fprime/discussions/new?category=pattern-request)

Also up-vote your faviorites: https://github.com/nasa/fprime/discussions/3298!
